File: data_structure/double_linked_list.cpp
#include <iostream>

class Node {
public:
    int data;
    Node* prev;
    Node* next;

    Node(int value) : data(value), prev(nullptr), next(nullptr) {}
};

class DoublyLinkedList {
private:
    Node* head;
    Node* tail;

public:
    DoublyLinkedList() : head(nullptr), tail(nullptr) {}

    // Append a new node to the end of the list
    void append(int data) {
        Node* newNode = new Node(data);
        if (!head) {
            head = tail = newNode;
        } else {
            tail->next = newNode;
            newNode->prev = tail;
            tail = newNode;
        }
    }

    // Prepend a new node to the beginning of the list
    void prepend(int data) {
        Node* newNode = new Node(data);
        if (!head) {
            head = tail = newNode;
        } else {
            newNode->next = head;
            head->prev = newNode;
            head = newNode;
        }
    }

    // Delete a node with a given value from the list
    void deleteNode(int data) {
        Node* current = head;
        while (current) {
            if (current->data == data) {
                if (current == head) {
                    head = current->next;
                    if (head) {
                        head->prev = nullptr;
                    }
                } else if (current == tail) {
                    tail = current->prev;
                    tail->next = nullptr;
                } else {
                    current->prev->next = current->next;
                    current->next->prev = current->prev;
                }
                delete current;
                return;
            }
            current = current->next;
        }
    }

    // Display the elements of the list
    void display() {
        Node* current = head;
        while (current) {
            std::cout << current->data << " <-> ";
            current = current->next;
        }
        std::cout << "nullptr" << std::endl;
    }
};

int main() {
    DoublyLinkedList dll;

    dll.append(1);
    dll.append(2);
    dll.append(3);
    dll.prepend(0);

    dll.display(); // Output: 0 <-> 1 <-> 2 <-> 3 <-> nullptr

    dll.deleteNode(2);

    dll.display(); // Output: 0 <-> 1 <-> 3 <-> nullptr

    return 0;
}

File: data_structure/double_linkedlist.py
class Node:
    def __init__(self,data):
        self.data=data
        self.next=None
        self.prev=None
class link:
    def __init__(self):
        self.head=None
    def insertBeg(self,newnode):
        if(self.head==None):
            self.head=newnode
        else:
            
            newnode.next=self.head
            self.head.prev=newnode
            self.head=newnode
    def insertEnd(self,newnode):
        if(self.head==None):
            self.head=newnode
        else:
            
            ptr=self.head
            while(ptr.next!=None):
                ptr=ptr.next
            ptr.next=newnode
            newnode.prev=ptr
    def insertAny(self,newnode,val):
        if(self.head==None):
            self.head=newnode
        else:
            
            ptr=self.head
            while(ptr.next!= None and ptr.data!=val):
                ptr=ptr.next
            newnode.next=ptr.next
            newnode.prev=ptr
            ptr.next.prev=newnode
            ptr.next=newnode
            
            
    def deleteBeg(self):
        strt=self.head
        self.head=strt.next
        self.head.prev=None
        del strt
    def deleteEnd(self):
        strt=self.head
        
        while(strt.next.next!=None):
            strt=strt.next
        ptr=strt.next
        strt.next=None
        del ptr
    def deleteAny(self,ele):
        strt=self.head
        
        while(strt.next!=None and strt.data!=ele):
            strt=strt.next
        ptr=strt
        strt.prev.next=strt.next
        strt.next.prev=strt.prev
        del ptr
    def print(self):
        nodes=self.head
        while(nodes!=None):
            print(nodes.data)
            nodes=nodes.next
n=int(input("Enter the number of nodes:"))
l=link()
for i in range (n):
    ele=int(input("enter the element in the list:"))
    enter=Node(ele)
    choice=int(input("enter 1 to enter at the beginning and 2 to add the end and 3 to enter at any position:"))
    match (choice):
           case 1:
             l.insertBeg(enter)
           case 2:
             l.insertEnd(enter)
           case 3:
             bef=int(input("Enter the element after which to be added:"))
             l.insertAny(enter,bef)
           case _:
             print("enter properly")
l.print()
d=int(input("Enter 1 to perform deletion and -1 to stop:"))
while(d!=-1):
    
    choice=int(input("enter 1 to delete at the beginning and 2 to delete the end and 3 to delete from any position:"))
    match (choice):
           case 1:
             l.deleteBeg()
           case 2:
             l.deleteEnd()
           case 3:
             bef=int(input("Enter the element to be deleted:"))
             l.deleteAny(enter,bef)
           case _:
             print("enter properly")
    d=int(input("Enter 1 to perform deletion and -1 to stop:"))
print("\n")
l.print()

        
